src/geometry/manifold/basic_smooth_bundle.lean

```diff
@@ -0,0 +1,594 @@
+        simp only [function.comp_app, model_with_corners_left_inv],
+        rw [j.1.left_inv],
+        exact hy.1.1.2 },
+      exact fderiv_within_congr U E (E _ M) },
+    have C : fderiv_within 𝕜 (I.to_fun ∘ u.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun)
+             (I.inv_fun ⁻¹' ((i.1.symm.trans j.1).trans (j.1.symm.trans u.1)).source ∩ range I.to_fun)
+             (I.to_fun x) =
+             fderiv_within 𝕜 (I.to_fun ∘ u.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun)
+             (range I.to_fun) (I.to_fun x),
+    { rw inter_comm,
+      apply fderiv_within_inter _ (I.unique_diff _ (mem_range_self _)),
+      apply I.continuous_inv_fun.continuous_at.preimage_mem_nhds
+        (mem_nhds_sets (local_homeomorph.open_source _) _),
+      simpa using hx },
+    have D : fderiv_within 𝕜 (I.to_fun ∘ u.1.to_fun ∘ j.1.inv_fun ∘ I.inv_fun)
+               (I.inv_fun ⁻¹' (j.1.symm.trans u.1).source ∩ range I.to_fun)
+               ((I.to_fun ∘ j.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun) (I.to_fun x)) =
+             fderiv_within 𝕜 (I.to_fun ∘ u.1.to_fun ∘ j.1.inv_fun ∘ I.inv_fun)
+               (range I.to_fun)
+               ((I.to_fun ∘ j.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun) (I.to_fun x)),
+    { rw inter_comm,
+      apply fderiv_within_inter _ (I.unique_diff _ (mem_range_self _)),
+      apply I.continuous_inv_fun.continuous_at.preimage_mem_nhds
+        (mem_nhds_sets (local_homeomorph.open_source _) _),
+      rw [local_homeomorph.trans_source] at hx,
+      simp [mem_inter_iff, -mem_image, -mem_range, mem_range_self],
+      exact hx.2 },
+    have E : fderiv_within 𝕜 (I.to_fun ∘ j.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun)
+               (I.inv_fun ⁻¹' ((i.1.symm.trans j.1).trans (j.1.symm.trans u.1)).source ∩ range I.to_fun)
+               (I.to_fun x) =
+             (fderiv_within 𝕜 (I.to_fun ∘ j.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun)
+               (range I.to_fun)
+               (I.to_fun x)),
+    { rw inter_comm,
+      apply fderiv_within_inter _ (I.unique_diff _ (mem_range_self _)),
+      apply I.continuous_inv_fun.continuous_at.preimage_mem_nhds
+        (mem_nhds_sets (local_homeomorph.open_source _) _),
+      simpa using hx },
+    rw [B, C, D, E] at A,
+    rw A,
+    assume v,
+    simp,
+    unfold_coes,
+    simp
+  end }
+
+/-- The tangent bundle to a smooth manifold, as a plain type. -/
+def tangent_bundle := (tangent_bundle_core I M).to_topological_fiber_bundle_core.total_space
+
+/-- The projection from the tangent bundle of a smooth manifold to the manifold. As the tangent
+bundle is represented internally as a product type, the notation `p.1` also works for the projection
+of the point `p`. -/
+def tangent_bundle.proj : tangent_bundle I M → M :=
+(tangent_bundle_core I M).to_topological_fiber_bundle_core.proj
+
+variable {M}
+
+/-- The tangent space at a point of the manifold `M`. It is just `E`. -/
+def tangent_space (x : M) : Type* :=
+(tangent_bundle_core I M).to_topological_fiber_bundle_core.fiber x
+
+section tangent_bundle_instances
```

Comment:
I think the intended instances would be much clearer if you could prove the instance term explicitly instead of setting reducibility and priorities so that `apply_instance` gets them. It would probably be more maintainable too.

Reply:
It is not really convenient to give the instances by hand: for example, the instance for `topological_space (tangent_space I x)` is `uniform_space.to_topological_space (tangent_space I x)`, where the uniform space structure is the one of `E` as `tangent_space I x` is just `E` as a type (and it is reducible now). Then the uniform space structure comes from the metric space structure, which itself comes from the normed space structure. And I don't want to register the normed space structure on `tangent_space I x`! It looks like the kind of situation where `apply_instance` is really doing a useful job.

I have added a comment to explain why one should change the priorities.